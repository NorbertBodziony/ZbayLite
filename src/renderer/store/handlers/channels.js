import Immutable from 'immutable'
import { DateTime } from 'luxon'
import { createAction, handleActions } from 'redux-actions'
import BigNumber from 'bignumber.js'

import {
  typeFulfilled,
  typeRejected,
  typePending,
  errorNotification,
  successNotification,
  LoaderState
} from './utils'
import notificationsHandlers from './notifications'
import channelHandlers from './channel'
import identitySelectors from '../selectors/identity'
import channelsSelectors from '../selectors/channels'
import modalsHandlers from './modals'
import { messages } from '../../zbay'
import { getVault } from '../../vault'
import { getClient } from '../../zcash'
import { networkFee } from '../../../shared/static'

const toBigNumber = x => new BigNumber(x)

export const ChannelsState = Immutable.Record(
  {
    data: Immutable.List(),
    loader: LoaderState({ loading: true })
  },
  'ChannelsState'
)

export const initialState = ChannelsState()

export const actionTypes = {
  LOAD_CHANNELS: 'LOAD_IDENTITY_CHANNELS',
  SET_LAST_SEEN: 'SET_CHANNELS_LAST_SEEN',
  SET_UNREAD: 'SET_CHANNEL_UNREAD'
}

const loadChannels = createAction(actionTypes.LOAD_CHANNELS, async id => {
  const channels = await getVault().channels.listChannels(id)

  return channels
})
const loadChannelsToNode = createAction(actionTypes.LOAD_CHANNELS, async id => {
  const channels = await getVault().channels.listChannels(id)
  await Promise.all(
    channels.map(channel =>
      getClient().keys.importIVK({ ivk: channel.keys.ivk, address: channel.address })
    )
  )
  await Promise.all(
    channels.filter(ch => ch.keys.sk).map(channel =>
      getClient().keys.importSK({ sk: channel.keys.sk, rescan: 'no' })
    )
  )
  return channels
})

const setLastSeen = createAction(actionTypes.SET_LAST_SEEN)
const setUnread = createAction(actionTypes.SET_UNREAD)

export const actions = {
  loadChannels,
  setLastSeen,
  setUnread,
  loadChannelsToNode
}

const _createChannel = async (identityId, { name, description }) => {
  const address = await getClient().addresses.create('sapling')
  const [ivk, sk] = await Promise.all([
    getClient().keys.exportIVK(address),
    getClient().keys.exportSK(address)
  ])
  const channel = {
    address,
    name: name,
    description: description,
    private: true,
    keys: {
      ivk,
      sk
    }
  }
  await getVault().channels.importChannel(identityId, channel)
  return address
}

const createChannel = (values, formActions) => async (dispatch, getState) => {
  const closeModal = modalsHandlers.actionCreators.closeModal('createChannel')
  const balance = identitySelectors.balance('zec')(getState())
  try {
    if (balance.lt(0.0002)) {
      dispatch(
        notificationsHandlers.actions.enqueueSnackbar(
          errorNotification({ message: `You need minimum 0.0002 ZEC to create a channel. ` })
        )
      )
      formActions.setSubmitting(false)
      return
    }
    const identityId = identitySelectors.id(getState())
    const address = await _createChannel(identityId, values)
    await dispatch(channelHandlers.epics.sendChannelSettingsMessage({ address: address }))
    dispatch(
      notificationsHandlers.actions.enqueueSnackbar(
        successNotification({ message: `Successfully created ${values.name} channel.` })
      )
    )
    formActions.setSubmitting(false)
    dispatch(closeModal())
    dispatch(loadChannels(identityId))
  } catch (error) {
    dispatch(
      notificationsHandlers.actions.enqueueSnackbar(
        errorNotification({ message: `Failed to create channel: ${error.message}` })
      )
    )
    formActions.setSubmitting(false)
  }
}

const withdrawMoneyFromChannels = () => async (dispatch, getState) => {
  const ownedChannels = channelsSelectors.ownedChannels(getState())
  let earnedAmount = toBigNumber(0)
  for (const channel of ownedChannels) {
    const channelAddress = channel.get('address')
    const amount = await getMoneyFromChannel(channelAddress)(dispatch, getState)
    earnedAmount = earnedAmount.plus(amount)
  }
  if (earnedAmount.gt(toBigNumber(0))) {
    dispatch(
      notificationsHandlers.actions.enqueueSnackbar(
        successNotification({
          message: `You will shortly receive ${earnedAmount.toFixed(
            4
          )} zash for commission from your created channels.`
        })
      )
    )
  }
}

const getMoneyFromChannel = address => async (dispatch, getState) => {
  const amount = await getClient().accounting.balance(address)
  const identityAddress = identitySelectors.address(getState())
  if (amount.gt(0.0005)) {
    const transfer = messages.createEmptyTransfer({
      address: identityAddress,
      amount: amount.minus(networkFee).toString(),
      identityAddress: address
    })
    await getClient().payment.send(transfer)
    return amount.minus(networkFee)
  }
  return toBigNumber(0)
}

const updateLastSeen = ({ channelId }) => async (dispatch, getState) => {
  const identity = identitySelectors.data(getState())
  const lastSeen = DateTime.utc()
  await getVault().channels.updateLastSeen({
    identityId: identity.get('id'),
    channelId,
    lastSeen
  })
  dispatch(setLastSeen({ channelId, lastSeen }))
  dispatch(setUnread({ channelId, unread: 0 }))
}

export const epics = {
  createChannel,
  updateLastSeen,
  getMoneyFromChannel,
  withdrawMoneyFromChannels
}

export const reducer = handleActions(
  {
    [typePending(actionTypes.LOAD_CHANNELS)]: state =>
      state.setIn(['loader', 'loading'], true).setIn(['loader', 'message'], 'Loading channels'),
    [typeFulfilled(actionTypes.LOAD_CHANNELS)]: (state, { payload: data }) =>
      state.set('data', Immutable.fromJS(data)).setIn(['loader', 'loading'], false),
    [typeRejected(actionTypes.LOAD_CHANNELS)]: (state, { payload: error }) =>
      state.setIn(['loader', 'loading'], false),
    [setLastSeen]: (state, { payload: { channelId, lastSeen } }) => {
      const index = state.data.findIndex(channel => channel.get('id') === channelId)
      return state.updateIn(['data', index], ch => ch.set('lastSeen', lastSeen))
    },
    [setUnread]: (state, { payload: { channelId, unread } }) => {
      const index = state.data.findIndex(channel => channel.get('id') === channelId)
      return state.updateIn(['data', index], ch => ch.set('unread', unread))
    }
  },
  initialState
)

export default {
  actions,
  epics,
  reducer
}
