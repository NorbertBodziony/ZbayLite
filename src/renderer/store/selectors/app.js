import { createSelector } from 'reselect'

const store = s => s

export const app = createSelector(store, state => state.get('app'))

const version = createSelector(app, a => a.version)

export default {
  version
}
