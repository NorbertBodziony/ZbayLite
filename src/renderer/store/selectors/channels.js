import { createSelector } from 'reselect'

const store = s => s

const channels = createSelector(store, state => state.get('channels'))
const data = createSelector(channels, ch => ch.data)
const loading = createSelector(channels, ch => ch.loading)
const errors = createSelector(channels, c => c.get('errors'))

export default {
  channels,
  loading,
  data,
  errors
}
