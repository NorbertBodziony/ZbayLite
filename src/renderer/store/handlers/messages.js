import Immutable from 'immutable'
import BigNumber from 'bignumber.js'
import * as R from 'ramda'
import { createAction, handleActions } from 'redux-actions'

import selectors from '../selectors/messages'
import appSelectors from '../selectors/app'
import channelsSelectors from '../selectors/channels'
import usersSelectors from '../selectors/users'
import identitySelectors from '../selectors/identity'
import operationsSelectors from '../selectors/operations'
import operationsHandlers from './operations'
import channelsHandlers from './channels'
import txnTimestampsHandlers from '../handlers/txnTimestamps'
import txnTimestampsSelector from '../selectors/txnTimestamps'
import appHandlers from './app'
import { messageType } from '../../zbay/messages'
import { messages as zbayMessages } from '../../zbay'
import { getClient } from '../../zcash'
import { displayMessageNotification } from '../../notifications'
import { getVault } from '../../vault'

export const MessageSender = Immutable.Record(
  {
    replyTo: '',
    username: ''
  },
  'MessageSender'
)

const _ReceivedMessage = Immutable.Record(
  {
    id: null,
    type: messageType.BASIC,
    sender: MessageSender(),
    createdAt: 0,
    message: '',
    spent: new BigNumber(0),
    isUnregistered: false,
    tag: '',
    offerOwner: ''
  },
  'ReceivedMessage'
)

export const ReceivedMessage = values => {
  const record = _ReceivedMessage(values)
  return record.set('sender', MessageSender(record.sender))
}

export const ChannelMessages = Immutable.Record(
  {
    messages: Immutable.List(),
    newMessages: Immutable.List()
  },
  'ChannelMessages'
)

export const initialState = Immutable.Map()

const setMessages = createAction('SET_MESSAGES')
const cleanNewMessages = createAction('CLEAN_NEW_MESSAGESS')
const appendNewMessages = createAction('APPEND_NEW_MESSAGES')

export const actions = {
  setMessages,
  cleanNewMessages,
  appendNewMessages
}

export const fetchMessages = channel => async (dispatch, getState) => {
  try {
    const pendingMessages = operationsSelectors.pendingMessages(getState())
    const identityAddress = identitySelectors.address(getState())
    const users = usersSelectors.users(getState())
    const txnTimestamps = txnTimestampsSelector.tnxTimestamps(getState())

    if (pendingMessages.find(msg => msg.status === 'pending')) {
      return
    }
    const channelId = channel.get('id')
    const previousMessages = selectors.currentChannelMessages(channelId)(getState())

    const transfers = await getClient().payment.received(channel.get('address'))

    if (transfers.length === appSelectors.transfers(getState()).get(channelId)) {
      return
    } else {
      dispatch(appHandlers.actions.setTransfers({ id: channelId, value: transfers.length }))
    }
    await transfers.forEach(async transfer => {
      if (!txnTimestamps.get(transfer.txid)) {
        const result = await getClient().confirmations.getResult(transfer.txid)
        await getVault().transactionsTimestamps.addTransaction(transfer.txid, result.blocktime)
        await dispatch(
          txnTimestampsHandlers.actions.addTxnTimestamp({
            tnxs: { [transfer.txid]: result.blocktime.toString() }
          })
        )
      }
    })
    const sortedTransfers = transfers.sort(
      (a, b) => txnTimestamps.get(a.txid) - txnTimestamps.get(b.txid)
    )
    const messagesAll = await Promise.all(
      sortedTransfers.map(async transfer => {
        const message = await zbayMessages.transferToMessage(transfer, users)
        if (message === null) {
          return ReceivedMessage(message)
        }
        const pendingMessage = pendingMessages.find(pm => pm.txId && pm.txId === message.id)
        if (pendingMessage) {
          dispatch(operationsHandlers.actions.removeOperation(pendingMessage.opId))
        }
        return ReceivedMessage(message)
      })
    )
    const messages = messagesAll.filter(message => message.id !== null)
    let lastSeen = channelsSelectors.lastSeen(channelId)(getState())
    if (!lastSeen) {
      await dispatch(channelsHandlers.epics.updateLastSeen({ channelId }))
      lastSeen = channelsSelectors.lastSeen(channelId)(getState())
    }
    const newMessages = zbayMessages.calculateDiff({
      previousMessages,
      nextMessages: Immutable.List(messages),
      lastSeen,
      identityAddress
    })
    await dispatch(channelsHandlers.actions.setUnread({ channelId, unread: newMessages.size }))
    await dispatch(
      appendNewMessages({
        channelId,
        messagesIds: newMessages.map(R.prop('id'))
      })
    )
    await dispatch(setMessages({ messages, channelId }))
    newMessages.map(nm => displayMessageNotification({ message: nm, channel }))
  } catch (err) {
    console.warn(err)
  }
}

export const epics = {
  fetchMessages
}

export const reducer = handleActions(
  {
    [setMessages]: (state, { payload: { channelId, messages } }) =>
      state.update(channelId, ChannelMessages(), cm =>
        cm.set('messages', Immutable.fromJS(messages))
      ),
    [cleanNewMessages]: (state, { payload: { channelId } }) =>
      state.update(channelId, ChannelMessages(), cm => cm.set('newMessages', Immutable.List())),
    [appendNewMessages]: (state, { payload: { channelId, messagesIds } }) =>
      state.update(channelId, ChannelMessages(), cm =>
        cm.update('newMessages', nm => nm.concat(messagesIds))
      )
  },
  initialState
)

export default {
  reducer,
  epics,
  actions
}
