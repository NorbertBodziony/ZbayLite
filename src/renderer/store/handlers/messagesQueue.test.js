/* eslint import/first: 0 */
jest.mock('../../zcash')

import Immutable from 'immutable'
import * as R from 'ramda'

import create from '../create'
import { actions, epics, initialState, DEFAULT_DEBOUNCE_INTERVAL } from './messagesQueue'
import selectors from '../selectors/messagesQueue'
import operationsSelectors from '../selectors/operations'
import notificationsSelectors from '../selectors/notifications'
import { createMessage, createChannel, now } from '../../testUtils'
import { ChannelsState } from './channels'
import { messageType } from '../../zbay/messages'
import { mock as zcashMock } from '../../zcash'

jest.useFakeTimers()

describe('Messages queue reducer handles', () => {
  let store = null

  beforeEach(() => {
    store = create({
      initialState: Immutable.Map({
        messagesQueue: initialState,
        channels: ChannelsState({
          data: R.range(0, 3).map(
            R.compose(
              Immutable.fromJS,
              createChannel
            )
          )
        })
      })
    })
    jest.clearAllMocks()
  })

  describe('actions', () => {
    describe('- addMessage', () => {
      it('when empty', () => {
        const message = createMessage('test-message-id')
        store.dispatch(actions.addMessage({ message, channelId: 'test-channel-id' }))
        expect(selectors.queue(store.getState())).toMatchSnapshot()
      })

      it('when has matching message', () => {
        const message = createMessage('test-message-id', now.toSeconds())
        const message2 = createMessage('test-message-id-2', now.plus({ seconds: 1 }).toSeconds())
        store.dispatch(actions.addMessage({ message, channelId: 'test-channel-id' }))
        store.dispatch(actions.addMessage({ message: message2, channelId: 'test-channel-id' }))
        expect(selectors.queue(store.getState())).toMatchSnapshot()
      })

      it('when different channel', () => {
        const message = createMessage('test-message-id', now.toSeconds())
        const message2 = createMessage('test-message-id-2', now.plus({ seconds: 1 }).toSeconds())
        store.dispatch(actions.addMessage({ message, channelId: 'test-channel-id' }))
        store.dispatch(actions.addMessage({ message: message2, channelId: 'test-channel-id-2' }))
        expect(selectors.queue(store.getState())).toMatchSnapshot()
      })

      it('when different type', () => {
        const message = createMessage('test-message-id', now.toSeconds())
        const message2 = createMessage('test-message-id-2', now.plus({ seconds: 1 }).toSeconds())
        store.dispatch(actions.addMessage({ message, channelId: 'test-channel-id' }))
        store.dispatch(actions.addMessage({
          message: {
            ...message2,
            type: messageType.AD
          },
          channelId: 'test-channel-id'
        }))
        expect(selectors.queue(store.getState())).toMatchSnapshot()
      })

      it('when different sender', () => {
        const message = createMessage('test-message-id', now.toSeconds())
        const message2 = createMessage('test-message-id-2', now.plus({ seconds: 1 }).toSeconds())
        store.dispatch(actions.addMessage({ message, channelId: 'test-channel-id' }))
        store.dispatch(actions.addMessage({
          message: {
            ...message2,
            sender: {
              replyTo: message2.sender.replyTo,
              username: 'Mercury'
            }
          },
          channelId: 'test-channel-id'
        }))
        expect(selectors.queue(store.getState())).toMatchSnapshot()
      })
    })

    it('- removeMessage', () => {
      const message = createMessage('test-message-id')
      const { payload: { key } } = store.dispatch(actions.addMessage({ message, channelId: 'test-channel-id' }))

      store.dispatch(actions.removeMessage(key))

      expect(selectors.queue(store.getState())).toMatchSnapshot()
    })
  })

  describe('epics', () => {
    describe('- sendPendingMessages', () => {
      beforeEach(() => {
        zcashMock.requestManager.z_getoperationstatus.mockImplementation(async () => [{
          id: 'operation-id',
          status: 'success',
          result: {
            txid: 'test-tx-id'
          },
          error: { code: -1, message: 'no error' }
        }])
      })

      it('sends messages', async () => {
        store.dispatch(actions.addMessage({
          message: createMessage('test-message-id', now.toSeconds()),
          channelId: 1
        }))
        store.dispatch(actions.addMessage({
          message: createMessage('test-message-id-2', now.plus({ seconds: 1 }).toSeconds()),
          channelId: 2
        }))

        const actionPromise = store.dispatch(epics.sendPendingMessages())
        jest.runAllTimers()
        await actionPromise

        expect(zcashMock.requestManager.z_sendmany.mock.calls).toMatchSnapshot()
      })

      it('adds operation observers', async () => {
        store.dispatch(actions.addMessage({
          message: createMessage('test-message-id', now.toSeconds()),
          channelId: 1
        }))
        store.dispatch(actions.addMessage({
          message: createMessage('test-message-id-2', now.plus({ seconds: 1 }).toSeconds()),
          channelId: 2
        }))
        zcashMock.requestManager.z_sendmany.mockResolvedValueOnce('op-id-1')
          .mockResolvedValueOnce('op-id-2')

        const actionPromise = store.dispatch(epics.sendPendingMessages())
        jest.runAllTimers()
        await actionPromise

        expect(operationsSelectors.operations(store.getState())).toMatchSnapshot()
      })

      it('clears queue', async () => {
        store.dispatch(actions.addMessage({
          message: createMessage('test-message-id', now.toSeconds()),
          channelId: 1
        }))
        store.dispatch(actions.addMessage({
          message: createMessage('test-message-id-2', now.plus({ seconds: 1 }).toSeconds()),
          channelId: 2
        }))

        const actionPromise = store.dispatch(epics.sendPendingMessages())
        jest.runAllTimers()
        await actionPromise

        expect(selectors.queue(store.getState())).toMatchSnapshot()
      })

      it('handles node errors using notifications', async () => {
        store.dispatch(actions.addMessage({
          message: createMessage('test-message-id', now.toSeconds()),
          channelId: 1
        }))
        zcashMock.requestManager.z_sendmany.mockRejectedValue('Test node error')

        const actionPromise = store.dispatch(epics.sendPendingMessages())
        jest.runAllTimers()
        await actionPromise

        const notifications = notificationsSelectors.data(store.getState())
        expect(notifications.map(n => n.remove('key'))).toMatchSnapshot()
      })
    })

    describe('- addMessage', () => {
      beforeEach(() => {
        zcashMock.requestManager.z_getoperationstatus.mockImplementation(async () => [{
          id: 'operation-id',
          status: 'success',
          result: {
            txid: 'test-tx-id'
          },
          error: { code: -1, message: 'no error' }
        }])
        zcashMock.requestManager.z_sendmany.mockResolvedValue('operation-id')
      })

      it('adds and sends message', async () => {
        const message = createMessage('test-message-id')

        const actionPromise = store.dispatch(epics.addMessage({ message, channelId: 1 }))
        jest.runAllTimers()
        await actionPromise

        expect(operationsSelectors.operations(store.getState())).toMatchSnapshot()
      })

      it('debounces sendPendingMessages', async () => {
        const message = createMessage('test-message-id')
        const message2 = createMessage('test-message-id-2')

        store.dispatch(epics.addMessage({ message, channelId: 1 }))
        jest.advanceTimersByTime(DEFAULT_DEBOUNCE_INTERVAL / 2)
        const actionPromise2 = store.dispatch(epics.addMessage({ message: message2, channelId: 1 }))
        jest.advanceTimersByTime(DEFAULT_DEBOUNCE_INTERVAL)
        await actionPromise2

        expect(operationsSelectors.operations(store.getState())).toMatchSnapshot()
      })

      it('does not debounce after interval', async () => {
        zcashMock.requestManager.z_sendmany.mockResolvedValueOnce('operation-id')
          .mockResolvedValueOnce('operation-id-2')
        const message = createMessage('test-message-id')
        const message2 = createMessage('test-message-id-2')

        const actionPromise = store.dispatch(epics.addMessage({ message, channelId: 1 }))
        jest.advanceTimersByTime(DEFAULT_DEBOUNCE_INTERVAL)
        await actionPromise
        const actionPromise2 = store.dispatch(epics.addMessage({ message: message2, channelId: 1 }))
        jest.advanceTimersByTime(DEFAULT_DEBOUNCE_INTERVAL)
        await actionPromise2

        expect(operationsSelectors.operations(store.getState())).toMatchSnapshot()
      })
    })
  })
})
