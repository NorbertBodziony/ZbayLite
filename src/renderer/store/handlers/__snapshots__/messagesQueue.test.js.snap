// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Messages queue reducer handles actions - addMessage when different channel 1`] = `
Immutable.Map {
  "81ef205fe0740b59c42077e13bcafae477437d737b748c7a4cce4026979cced8": Immutable.PendingMessage {
    "channelId": "test-channel-id",
    "message": Immutable.Map {
      "id": "test-message-id",
      "type": 1,
      "createdAt": 1551963828,
      "message": "This is some message test-message-id",
      "sender": Immutable.Map {
        "replyTo": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
      },
    },
  },
  "b211c5afc2ce5af6be220b354ccd971a6af720f4d15581cf3e4b3b59aac00997": Immutable.PendingMessage {
    "channelId": "test-channel-id-2",
    "message": Immutable.Map {
      "id": "test-message-id-2",
      "type": 1,
      "createdAt": 1551963829,
      "message": "This is some message test-message-id-2",
      "sender": Immutable.Map {
        "replyTo": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
      },
    },
  },
}
`;

exports[`Messages queue reducer handles actions - addMessage when different sender 1`] = `
Immutable.Map {
  "81ef205fe0740b59c42077e13bcafae477437d737b748c7a4cce4026979cced8": Immutable.PendingMessage {
    "channelId": "test-channel-id",
    "message": Immutable.Map {
      "id": "test-message-id",
      "type": 1,
      "createdAt": 1551963828,
      "message": "This is some message test-message-id",
      "sender": Immutable.Map {
        "replyTo": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
      },
    },
  },
  "27954fbac6e2e9d985a252eabdb605852865222d62dde2a2a71df48199a0f54e": Immutable.PendingMessage {
    "channelId": "test-channel-id",
    "message": Immutable.Map {
      "id": "test-message-id-2",
      "type": 1,
      "createdAt": 1551963829,
      "message": "This is some message test-message-id-2",
      "sender": Immutable.Map {
        "replyTo": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
        "username": "Mercury",
      },
    },
  },
}
`;

exports[`Messages queue reducer handles actions - addMessage when different type 1`] = `
Immutable.Map {
  "81ef205fe0740b59c42077e13bcafae477437d737b748c7a4cce4026979cced8": Immutable.PendingMessage {
    "channelId": "test-channel-id",
    "message": Immutable.Map {
      "id": "test-message-id",
      "type": 1,
      "createdAt": 1551963828,
      "message": "This is some message test-message-id",
      "sender": Immutable.Map {
        "replyTo": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
      },
    },
  },
  "c547dbe9e867e012bb11db7f4a8f1f4680e7780513901f22c2d166bda0a28c4d": Immutable.PendingMessage {
    "channelId": "test-channel-id",
    "message": Immutable.Map {
      "id": "test-message-id-2",
      "type": 2,
      "createdAt": 1551963829,
      "message": "This is some message test-message-id-2",
      "sender": Immutable.Map {
        "replyTo": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
      },
    },
  },
}
`;

exports[`Messages queue reducer handles actions - addMessage when empty 1`] = `
Immutable.Map {
  "81ef205fe0740b59c42077e13bcafae477437d737b748c7a4cce4026979cced8": Immutable.PendingMessage {
    "channelId": "test-channel-id",
    "message": Immutable.Map {
      "id": "test-message-id",
      "type": 1,
      "createdAt": 1551963828,
      "message": "This is some message test-message-id",
      "sender": Immutable.Map {
        "replyTo": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
      },
    },
  },
}
`;

exports[`Messages queue reducer handles actions - addMessage when has matching message 1`] = `
Immutable.Map {
  "81ef205fe0740b59c42077e13bcafae477437d737b748c7a4cce4026979cced8": Immutable.PendingMessage {
    "channelId": "test-channel-id",
    "message": Immutable.Map {
      "id": "test-message-id",
      "type": 1,
      "createdAt": 1551963828,
      "message": "This is some message test-message-id
This is some message test-message-id-2",
      "sender": Immutable.Map {
        "replyTo": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
      },
    },
  },
}
`;

exports[`Messages queue reducer handles actions - removeMessage 1`] = `Immutable.Map {}`;

exports[`Messages queue reducer handles epics - addMessage adds and sends message 1`] = `
Immutable.Map {
  "operation-id": Immutable.Operation {
    "opId": "operation-id",
    "type": "pendingMessage",
    "meta": Immutable.Record {
      "message": Immutable.Map {
        "id": "test-message-id",
        "type": 1,
        "createdAt": 1551963828,
        "message": "This is some message test-message-id",
        "sender": Immutable.Map {
          "replyTo": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
        },
      },
      "channelId": 1,
    },
    "txId": "test-tx-id",
    "error": Immutable.ZcashError {
      "code": -1,
      "message": "no error",
    },
    "status": "success",
  },
}
`;

exports[`Messages queue reducer handles epics - addMessage debounces sendPendingMessages 1`] = `
Immutable.Map {
  "operation-id": Immutable.Operation {
    "opId": "operation-id",
    "type": "pendingMessage",
    "meta": Immutable.Record {
      "message": Immutable.Map {
        "id": "test-message-id",
        "type": 1,
        "createdAt": 1551963828,
        "message": "This is some message test-message-id
This is some message test-message-id-2",
        "sender": Immutable.Map {
          "replyTo": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
        },
      },
      "channelId": 1,
    },
    "txId": "test-tx-id",
    "error": Immutable.ZcashError {
      "code": -1,
      "message": "no error",
    },
    "status": "success",
  },
}
`;

exports[`Messages queue reducer handles epics - addMessage does not debounce after interval 1`] = `
Immutable.Map {
  "operation-id": Immutable.Operation {
    "opId": "operation-id",
    "type": "pendingMessage",
    "meta": Immutable.Record {
      "message": Immutable.Map {
        "id": "test-message-id",
        "type": 1,
        "createdAt": 1551963828,
        "message": "This is some message test-message-id",
        "sender": Immutable.Map {
          "replyTo": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
        },
      },
      "channelId": 1,
    },
    "txId": "test-tx-id",
    "error": Immutable.ZcashError {
      "code": -1,
      "message": "no error",
    },
    "status": "success",
  },
  "operation-id-2": Immutable.Operation {
    "opId": "operation-id-2",
    "type": "pendingMessage",
    "meta": Immutable.Record {
      "message": Immutable.Map {
        "id": "test-message-id-2",
        "type": 1,
        "createdAt": 1551963828,
        "message": "This is some message test-message-id-2",
        "sender": Immutable.Map {
          "replyTo": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
        },
      },
      "channelId": 1,
    },
    "txId": "test-tx-id",
    "error": Immutable.ZcashError {
      "code": -1,
      "message": "no error",
    },
    "status": "success",
  },
}
`;

exports[`Messages queue reducer handles epics - sendPendingMessages adds operation observers 1`] = `
Immutable.Map {
  "op-id-1": Immutable.Operation {
    "opId": "op-id-1",
    "type": "pendingMessage",
    "meta": Immutable.Record {
      "message": Immutable.Map {
        "id": "test-message-id",
        "type": 1,
        "createdAt": 1551963828,
        "message": "This is some message test-message-id",
        "sender": Immutable.Map {
          "replyTo": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
        },
      },
      "channelId": 1,
    },
    "txId": "test-tx-id",
    "error": Immutable.ZcashError {
      "code": -1,
      "message": "no error",
    },
    "status": "success",
  },
  "op-id-2": Immutable.Operation {
    "opId": "op-id-2",
    "type": "pendingMessage",
    "meta": Immutable.Record {
      "message": Immutable.Map {
        "id": "test-message-id-2",
        "type": 1,
        "createdAt": 1551963829,
        "message": "This is some message test-message-id-2",
        "sender": Immutable.Map {
          "replyTo": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
        },
      },
      "channelId": 2,
    },
    "txId": "test-tx-id",
    "error": Immutable.ZcashError {
      "code": -1,
      "message": "no error",
    },
    "status": "success",
  },
}
`;

exports[`Messages queue reducer handles epics - sendPendingMessages clears queue 1`] = `Immutable.Map {}`;

exports[`Messages queue reducer handles epics - sendPendingMessages handles node errors using notifications 1`] = `
Immutable.List [
  Immutable.Map {
    "options": Immutable.Map {
      "persist": true,
      "variant": "error",
      "action": [Function],
    },
    "message": "Couldn't send the message, please check node connection.",
  },
]
`;

exports[`Messages queue reducer handles epics - sendPendingMessages sends messages 1`] = `
Array [
  Array [
    "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
    Array [
      Object {
        "address": "zs1testaddress1",
        "amount": "0.0001",
        "memo": "1787997c30e94f050c634d9dbe9b948e6bccbf9fc0dcfbd4d441e5f7d57db3db1deebe6357dbf5601fed3400000000000000000000000000000000000000005c8116b401654a785443736e494c4659416f754c3833465346334e54693473543056495753314f4953585368484e7a4e4643514148536732650000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
      },
    ],
    1,
    0.0001,
  ],
  Array [
    "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
    Array [
      Object {
        "address": "zs1testaddress2",
        "amount": "0.0001",
        "memo": "1787997c30e94f050c634d9dbe9b948e6bccbf9fc0dcfbd4d441e5f7d57db3db1deebe6357dbf5601fed3400000000000000000000000000000000000000005c8116b501654a785443736e494c4659416f754c3833465346334e54693473543056495753314f4953585368484e7a4e463130674a41434d744466303d00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
      },
    ],
    1,
    0.0001,
  ],
]
`;
